[mlir][tensor] Fix crash when canonicalizing invalid IR

matthias-springer opened the pull request:
This commit fixes a crash of the canonicalizer when there are slice ops with offset/size SSA values that have a negative constant value. Such ops are invalid if they are reachable and their offsets/sizes should not be folded to static integer values. (But such ops may appear in non-reachable block.)

This commit fixes #71150.

joker-eph commented on mlir/lib/Dialect/Utils/StaticValueUtils.cpp:
```diff
@@ -256,13 +256,20 @@ std::optional<int64_t> constantTripCount(OpFoldResult lb, OpFoldResult ub,
+      if (intVal.isNegative() && onlyNonNegative)
+        continue;
+      Attribute attr;
+      bool isConstant = matchPattern(ofr.get<Value>(), m_Constant(&attr));
```
Is this the most efficient way? (knowing that you just matched in the `if` above)

matthias-springer committed:
[mlir][tensor] Fix crash when canonicalizing invalid IR

This commit fixes a crash of the canonicalizer when there are slice ops with offset/size SSA values that have a negative constant value. Such ops are invalid if they are reachable and their offsets/sizes should not be folded to static integer values. (But such ops may appear in non-reachable block.)

This commit partially fixes #71150. The canonicalizer no longer crashes, but invalid IR is still being produced.
